src/ResourceManager/Common/Commands.ResourceManager.Common/Location/LocationCompleter.cs

```diff
@@ -0,0 +1,200 @@
+﻿// ----------------------------------------------------------------------------------
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.ResourceManager.Common.Location
+{
+    using Commands.Common.Authentication.Abstractions;
+    using Commands.Common.Authentication;
+    using Internal.Subscriptions;
+    using Properties;
+    using Management.Internal.Resources.Models;
+    using Management.Internal.Resources;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Management.Automation;
+    using Rest;
+    using Commands.Common.Authentication.Models;
+
+
+    /// <summary>
+    /// This attribute will allow the user to autocomplete the -Location parameter of a cmdlet with valid locations (as determined by the list of ResourceTypes given)
+    /// </summary>
+    public class LocationCompleterAttribute : ArgumentCompleterAttribute
+    {
+        private static IDictionary<int, IDictionary<string, ICollection<string>>> _resourceTypeLocationDictionary = new Dictionary<int, IDictionary<string, ICollection<string>>>();
+
+        private static readonly object _lock = new object();
+
+        protected static IDictionary<string, ICollection<string>> ResourceTypeLocationDictionary
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    IAzureContext context = AzureRmProfileProvider.Instance.Profile.DefaultContext;
+
+                    var contextHash = HashContext(context);
+
+                    if (!_resourceTypeLocationDictionary.ContainsKey(contextHash))
+                    {
+                        try
+                        {
+                            IResourceManagementClient client = AzureSession.Instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
```

Comment:
Why create a custom arm client, why not just use CreateArmClient, and provide the context and endpoint name?

Reply:
@maddieclayton Mark is referring to something like [this](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/SdkClient/ResourceManagerSdkClient.cs#L81-L90)

Reply:
I'm not sure how that implementation works, because when it runs I get the error "Cannot create instance of management client type ResourceManagementClient. It does not have the expected constructor."  I'll discuss it more with Cormac when he gets back.